src/Nethermind/Nethermind.Blockchain/Blocks/BlockStore.cs

```diff
@@ -51,33 +52,58 @@ public void Insert(Block block)
         // if we carry Rlp from the network message all the way here then we could solve 4GB of allocations and some processing
         // by avoiding encoding back to RLP here (allocations measured on a sample 3M blocks Goerli fast sync
         using NettyRlpStream newRlp = _blockDecoder.EncodeToNewNettyStream(block);
-        _blockDb.Set(block.Hash, newRlp.AsSpan());
+
+        _blockDb.Set(block.Number, block.Hash, newRlp.AsSpan());
     }
 
-    public void Delete(Keccak blockHash)
+    private static void GetBlockNumPrefixedKey(long blockNumber, Keccak blockHash, Span<byte> output)
     {
-        _blockDb.Delete(blockHash);
-        _blockCache.Delete(blockHash);
+        blockNumber.WriteBigEndian(output);
+        blockHash!.Bytes.CopyTo(output[8..]);
     }
 
-    public Block? Get(Keccak blockHash, bool shouldCache = false)
+    public void Delete(long blockNumber, Keccak blockHash)
     {
+        _blockDb.Remove(blockHash.Bytes);
+        _blockDb.Delete(blockNumber, blockHash);
+        _blockCache.Delete(blockHash);
```

Comment:
reverse order of removal, first cache, then mapping, then data -> cannot corrupt anything even if something breaks.

Reply:
Done